renku_notebooks/api/notebooks.py

```diff
@@ -79,6 +79,7 @@ def notebook_status(user, namespace, project, commit_sha, notebook=None):
 @authenticated
 def launch_notebook(user, namespace, project, commit_sha, notebook=None):
     """Launch user server with a given name."""
+    branch = request.args.get("branch", "master")
```

Comment:
shouldn't the route at line 77/78 be updated to something like `"<namespace>/<project>/<branch>/<commit_sha>"` to keep things coherent?

Reply:
A branch is not a unique property while a commit is (e.g. you can have two branches on the exact same commit). That's why it was excluded to begin with. What about passing the commit *and* the branch as data? E.g. you would make a post to `<namespace>/<project>` and specify a json with 

```
{
  "commit-sha": "12345",
  "ref": "some-branch"
}
```

Reply:
The `commit-sha` would be required and `ref` optional.

Reply:
Everywhere in the notebook service with use the combination of `<namespace>/<project>/<commit_sha>` to uniquely identify an endpoint. If we spawn servers from different branches then we need `<branch>` to be able to do that (e.g. two different servers from two different branches on the same commit should be identifiable for DELETE). So I prefer to have it in the endpoint url as @lorenzo-cavazzi suggested.

Reply:
Moreover, I prefer to stick to `branch` and not `ref` to avoid confusion.

Reply:
Also, if you don't mind I'll include your changes in my PR and update the notebook services API as well.

Reply:
but then it should be `<namespace>/<project>/<commit-sha>/<branch>` since you can have multiple branches per commit. 

What should a `GET` return? I suppose on `<namespace>/<project>` it should return a list of servers running per commit; on `<namespace>/<project>/<commit-sha>` it should return the list of servers identified by branch for that commit, and on `<namespace>/<project>/<commit-sha>/<branch>` it should return the status of the uniquely-identified server.

Reply:
> Also, if you don't mind I'll include your changes in my PR and update the notebook services API as well.

why? this PR has nothing to do with refactoring the other endpoint.

Reply:
from a rest api point of view, I don't really like the URL sprawl - a POST on the `<namespace>/<project>/<commit>` endpoint doesn't make sense because we cannot create any objects there. The object we are creating with a POST is a server for the project. That's why I like the idea of having a single endpoint be `<namespace>/<project>` (or with `/servers`) at the end - a GET retrieves the list of running servers, a POST creates a new one. A GET with `commit` and `branch` in the payload gets you the specific server.